jupyter_server_proxy/handlers.py

```diff
@@ -54,6 +54,20 @@ def _default_code(self):
     def _default_reason(self):
         return self.orig_response.reason
 
+    @observe('code')
+    def _observe_code(self, change):
+        # Automatically update reason if code changes and previous reason was default.
+        old_default_reason = httputil.responses.get(change['old'], 'Unknown')
+        new_default_reason = httputil.responses.get(change['new'], 'Unknown')
+        if self.reason == old_default_reason:
+            self.reason = new_default_reason
```

Comment:
Is this really correct?

I think `change["old"]` reflects the old value of the `code` propery, and `change["new"]` reflects the new value of the code property. Based on the variable names, it seems like they are populated with the wrong content, and therefore the if statement below does something weird as well.

I think as part of a function observing `code`, you can only reason about before/after values of `code`, not before/after values of reason.

It seems to me like the variables are assigned a old / new value of the `code` propery, and that this is compared to the `reason` property. If you have logic comparing something with reason, shouldn't it be a comparison between reasons?

Reply:
Thanks @consideRatio for taking a look!

> Based on the variable names, it seems like they are populated with the wrong content

Hmm, it looks correct to me, I don't understand what you mean. Could you perhaps spell it out more explicitly?

I'll write out an example here of how this logic is supposed to work. Hopefully that will eliminate confusion...

Reply:
Let's say the initial response has `code=200` and `reason="OK"`. If I rewrite the response so that `code=404`, then I want `reason` to automatically switch to `404` as long as `reason` was not modified. In this example,

```python
change['old'] = 200
old_default_reason = "OK"
change['new'] = 404
new_default_reason = "Not Found"
```
Then the if statement checks that `self.reason == "OK"` and adjusts it to `self.reason = "Not Found"`.

The effect is that `reason` remains consistent if one only adjusts `code`. But one can still adjust `reason` independently.

Does this make sense? Is my code actually consistent with what I'm saying? (Maybe I'm blind to some typo I made.)

Reply:
If you want to set reason to the response code, then you are doing it - practically. But, consider for example...

```
        old_default_reason = httputil.responses.get(change['old'], 'Unknown')
```

Isn't this really just setting a variable named `old_default_reason` to the old response code? Shouldn't you name the variable `old_response_code` then and let your logic act based on that?

I think you are making an implicit assumption that the response reason is the response code, but it seems weird for me that you would have it like that.

Reply:
> It seems to me, that you don't want the reason to be the same as the code?

Well, `code` is an integer, and `reason` is a textual description which should correspond to that code. (And `httputil.responses` is a dictionary mapping numerical codes to textual descriptions.) Since `code`s are `int`s and `reason`s are `str`s, I don't understand how they could be the same.

I do want the reason to _correspond_ to the code, unless a custom string is being provided for `reason`.

Perhaps another example would help? I want to be able to do

```python
response.reason = "The server is sleepy"
response.code = 500
```
or equivalently
```python
response.code = 500
response.reason = "The server is sleepy"
```

I agree that things are confusing, although I don't yet understand your particular confusion. I think the jsp code would greatly benefit from type annotations. :smile:

Reply:
> (And httputil.responses is a dictionary mapping numerical codes to textual descriptions.)

This was my confusion, I'm not sure what I thought... I now understand it very differently!

Is it correct you want reason to be based on the code, and kept updated based on the code, unless someone has explicitly set the reason?

Reply:
Based on the comment "Automatically update reason if code changes and previous reason was default.", I think this is the code I'd write.

```python
    def _observe_code(self, change):
        # Automatically update reason if code changes and previous reason was default.
        if self.reason == self._default_reason():
            self.reason = httputil.responses.get(change['new'], 'Unknown')
```

But, I think you may be looking for logic to automatically update the reason if a user has explicitly set it in a way that isn't generated based on the code? Hmmm... I would try write this a bit differently. Maybe this would work for example, and then you remove the entire `def _observe_code(self, change):` function?

```python
    @default('reason')
    def _default_reason(self):
        return httputil.responses.get(self.code, 'Unknown')
```

Reply:
What about...

```suggestion
        # HTTP status codes are mapped to short descriptions in the
        # httputil.responses dictionary, 200 maps to "OK", 403 maps to
        # "Forbidden" etc.
        #
        # If code is updated and it previously had a reason matching its short
        # description, we update reason to match the new code's short
        # description.
        #
        if self.reason == httputil.responses.get(change['old'], 'Unknown'):
            self.reason = httputil.responses.get(change['new'], 'Unknown')
```